content/style-injector.js

```diff
@@ -1,184 +1,335 @@
-      el = document.createElement('style');
+      el = createElement.call(creationDoc, 'style');
+    }
+    if (id) {
+      el.id = `${PREFIX}${id}`;
+      const oldEl = document.getElementById(el.id);
+      if (oldEl) oldEl.id += '-superseded-by-Stylus';
     }
-    el.id = `${PREFIX}${id}`;
     el.type = 'text/css';
     // SVG className is not a string, but an instance of SVGAnimatedString
     el.classList.add('stylus');
+    el.textContent = code;
     return el;
   }
 
-  function clear() {
-    for (const style of list) {
-      style.el.remove();
+  function _emitUpdate() {
+    if (!IS_OWN_PAGE && list.length) {
+      docRewriteObserver.start();
+      docRootObserver.start();
+    } else {
+      docRewriteObserver.stop();
+      docRootObserver.stop();
     }
-    list.length = 0;
-    table.clear();
-    emitUpdate();
+    onUpdate();
   }
 
-  function toggle(_enabled) {
-    if (enabled === _enabled) return;
-    enabled = _enabled;
-    for (const style of list) {
-      style.el.disabled = !enabled;
+  /*
+  FF59+ workaround: allow the page to read our sheets, https://github.com/openstyles/stylus/issues/461
+  First we're trying the page context document where inline styles may be forbidden by CSP
+  https://bugzilla.mozilla.org/show_bug.cgi?id=1579345#c3
+  and since userAgent.navigator can be spoofed via about:config or devtools,
+  we're checking for getPreventDefault that was removed in FF59
+  */
+  function _initCreationDoc() {
+    creationDoc = !Event.prototype.getPreventDefault && document.wrappedJSObject;
+    if (creationDoc) {
+      ({createElement, createElementNS} = creationDoc);
+      const el = document.documentElement.appendChild(_createStyle());
+      const isApplied = el.sheet;
+      el.remove();
+      if (isApplied) return;
     }
+    creationDoc = document;
+    ({createElement, createElementNS} = document);
   }
 
-  function sort() {
-    list.sort(compare);
-    for (const style of list) {
-      // FIXME: do we need this?
-      // const copy = document.importNode(el, true);
-      // el.textContent += ' '; // invalidate CSSOM cache
-      document.documentElement.appendChild(style.el);
-      // moving an element resets its 'disabled' state
-      style.el.disabled = !enabled;
+  function _remove(id) {
+    const style = table.get(id);
+    if (!style) return;
+    table.delete(id);
+    list.splice(list.indexOf(style), 1);
+    style.el.remove();
+  }
+
+  function _sort() {
+    docRootObserver.evade(() => {
+      list.sort(compare);
+      for (const style of list) {
+        // moving an element resets its 'disabled' state
+        document.documentElement.appendChild(style.el);
+        style.el.disabled = !isEnabled;
+      }
+    });
+  }
+
+  function _sortIfNeeded() {
+    if (!list.length) {
+      return false;
+    }
+    let el = list[0].el;
+    if (el.parentNode !== creationDoc.documentElement) {
+      return true;
```

Comment:
It seems that this branch is handled by `RewriteObserver`. Is it safe to remove?

Reply:
My bad, it's totally wrong. I didn't notice there were multiple returns so now I rewrote it to a single exit point which should take care of the problem, if I understand your question correctly.